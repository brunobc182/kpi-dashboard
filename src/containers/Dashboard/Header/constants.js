import { COLORS } from "../../../theme";

/**
 * Represents the card title and bgColor props
 *
 * @constant
 * @type {object}.
 */
export const CARD_CONFIG = {
  active_source: {
    bgColor: `linear-gradient(120deg, ${COLORS.BRINK_PINK} 0%, ${COLORS.POMEGRANATE} 40%)`,

    title: "Active sourcing",
  },
  weekly_active: {
    bgColor: `linear-gradient(120deg, ${COLORS.CORNFLOWER_BLUE} 0%, ${COLORS.BLUE_RIBBON} 40%)`,
    title: "Weekly active",
  },
  nps: {
    bgColor: `linear-gradient(120deg, ${COLORS.PASTEL_GREEN} 0%, ${COLORS.JADE} 40%)`,
    title: "NPS",
  },
};

/**
 * Represents the filter optins
 *
 * @constant
 * @type {object}.
 */

export const FILTER_OPTIONS = [
  {
    time_unit: "WEEK",
    time_unit_count: 1,
  },
  {
    time_unit: "MONTH",
    time_unit_count: 1,
  },
  {
    time_unit: "QUARTER",
    time_unit_count: 1,
  },
  {
    time_unit: "YEAR",
    time_unit_count: 1,
  },
];
