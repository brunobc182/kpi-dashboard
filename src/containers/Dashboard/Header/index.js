import React from "react";
import styled from "styled-components";
import axios from "axios";
import { Card, Filter } from "../../../components";
import { CARD_CONFIG, FILTER_OPTIONS } from "./constants";
import { COLORS, DEVICE } from "../../../theme";

const Container = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  max-width: 1200px;
  margin: 0 auto;
  padding: 30px;
`;

const FilterWrapper = styled.div`
  display: flex;
  align-self: center;

  @media ${DEVICE.TABLET} {
    align-self: flex-end;
  }
`;

const CardsWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;

  @media ${DEVICE.TABLET} {
    flex-direction: row;
  }
`;

const Loading = styled.span`
  font-weight: bold;
`;

const ErrorMessage = styled.span`
  color: ${COLORS.SCARLET};
`;

function parseFilterOptions(options) {
  let tempOptions = [];

  options.forEach((option) => {
    tempOptions.push({
      value: option.time_unit,
      text: option.time_unit,
    });
  });

  return tempOptions;
}

function Header() {
  const [selectectFilter, setSelectedFilter] = React.useState("MONTH");
  const [cardsData, setCardsData] = React.useState(null);
  const [cardsListError, setCardsListError] = React.useState(false);
  const [isLoading, setIsLoading] = React.useState(false);
  const [requestPayload, setRequestPayload] = React.useState({
    time_unit: "MONTH",
    time_unit_count: 1,
  });

  const filterOptions = parseFilterOptions(FILTER_OPTIONS);

  function handleFilterChange(event) {
    const currentFilter = event.target.value;
    const currentPayload = FILTER_OPTIONS.filter(
      (option) => currentFilter === option.time_unit
    );

    setSelectedFilter(currentFilter);
    setRequestPayload(...currentPayload);
  }

  async function fetchData(requestPayload) {
    setIsLoading(true);

    try {
      const response = await axios.get("data.json");

      setIsLoading(false);

      if (response?.data.status !== "ok") {
        setCardsListError(true);
      }

      setCardsData(response.data.data);
    } catch (error) {
      setIsLoading(false);
      setCardsListError(true);
    }
  }

  React.useEffect(() => {
    fetchData(requestPayload);
  }, [requestPayload]);

  if (isLoading) {
    return (
      <Container>
        <Loading data-testid="loading">Loading data...</Loading>
      </Container>
    );
  }

  return (
    <Container>
      <FilterWrapper>
        <Filter
          testeId="filter"
          selectectValue={selectectFilter}
          options={filterOptions}
          handleChange={handleFilterChange}
        />
      </FilterWrapper>
      {cardsListError ? (
        <ErrorMessage data-testid="error">
          An error occurred when fetch the cards list, please try again.
        </ErrorMessage>
      ) : (
        <CardsWrapper data-testid="cards">
          {cardsData &&
            Object.keys(cardsData).map((card) => (
              <Card
                bgColor={CARD_CONFIG[card].bgColor}
                key={card}
                title={CARD_CONFIG[card].title}
                currentPeriod={cardsData[card].current_period}
                lastPeriod={cardsData[card].last_period}
              />
            ))}
        </CardsWrapper>
      )}
    </Container>
  );
}

export default Header;
