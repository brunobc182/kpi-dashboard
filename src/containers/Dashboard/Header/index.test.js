import "@testing-library/jest-dom";
import React from "react";
import { render, cleanup, waitFor, fireEvent } from "@testing-library/react";
import { act } from "react-dom/test-utils";

import Header from "./";
import axiosMock from "axios";

afterEach(() => {
  cleanup();
  jest.clearAllMocks();
});

describe("Header", () => {
  it("fetches data and renders correctly", async () => {
    axiosMock.get.mockResolvedValueOnce({
      data: {
        data: {
          active_source: {
            current_period: 45,
            last_period: 30,
          },
        },
        status: "ok",
      },
    });

    const { container, getByTestId } = render(<Header />);
    expect(getByTestId("loading")).toHaveTextContent(/^Loading data...$/);
    expect(container).toMatchSnapshot();

    const cards = await waitFor(() => getByTestId("cards"));

    expect(cards).toBeInTheDocument();
    expect(axiosMock.get).toHaveBeenCalledTimes(1);
    expect(axiosMock.get).toHaveBeenCalledWith("data.json");
    waitFor(() => expect(container).toMatchSnapshot());
  });

  it("shows an error message when fetch data with a failed status", async () => {
    axiosMock.get.mockResolvedValueOnce({
      data: {
        data: {},
        status: "error",
      },
    });

    const { container, getByTestId } = render(<Header />);
    const erroMessage = await waitFor(() => getByTestId("error"));

    expect(erroMessage).toHaveTextContent(
      /^An error occurred when fetch the cards list, please try again.$/
    );
    waitFor(() => expect(container).toMatchSnapshot());
  });

  it("shows an error message when promise gets rejected", async () => {
    axiosMock.get.mockRejectedValueOnce({});

    const { container, getByTestId } = render(<Header />);
    const erroMessage = await waitFor(() => getByTestId("error"));

    expect(erroMessage).toHaveTextContent(
      /^An error occurred when fetch the cards list, please try again.$/
    );
    waitFor(() => expect(container).toMatchSnapshot());
  });

  it("fetches new data when a filter is selected", async () => {
    axiosMock.get.mockResolvedValue({
      data: {
        data: {
          active_source: {
            current_period: 45,
            last_period: 30,
          },
        },
        status: "ok",
      },
    });

    await act(async () => {
      const { getByTestId } = render(<Header />);
      const filter = await waitFor(() => getByTestId("filter"));
      fireEvent.change(filter, { target: { value: "WEEK" } });
    });

    expect(axiosMock.get).toHaveBeenCalledTimes(2);
    expect(axiosMock.get).toHaveBeenCalledWith("data.json");
  });
});
