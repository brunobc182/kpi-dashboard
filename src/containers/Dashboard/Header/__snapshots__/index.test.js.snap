// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header fetches data and renders correctly 1`] = `
<div>
  <div
    class="sc-AxmLO iyRexG"
  >
    <span
      class="sc-fzpans iEVhKc"
      data-testid="loading"
    >
      Loading data...
    </span>
  </div>
</div>
`;

exports[`Header fetches data and renders correctly 2`] = `
<div>
  <div
    class="sc-AxmLO iyRexG"
  >
    <div
      class="sc-fzozJi csyUvS"
    >
      <div
        class="sc-AxhUy cuOvzs"
      >
        <label
          class="sc-AxgMl fxzWMk"
        >
          Filter by
        </label>
        <select
          class="sc-AxheI dXBtZm"
          data-testid="filter"
        >
          <option
            class="sc-Axmtr hBVTto"
            value="WEEK"
          >
            WEEK
          </option>
          <option
            class="sc-Axmtr hBVTto"
            value="MONTH"
          >
            MONTH
          </option>
          <option
            class="sc-Axmtr hBVTto"
            value="QUARTER"
          >
            QUARTER
          </option>
          <option
            class="sc-Axmtr hBVTto"
            value="YEAR"
          >
            YEAR
          </option>
        </select>
      </div>
    </div>
    <div
      class="sc-fzoLsD kiZmwC"
      data-testid="cards"
    >
      <div
        class="sc-AxjAm chaqBE"
      >
        <h4
          class="sc-AxirZ jfgKYb"
        >
          Active sourcing
        </h4>
        <p
          class="sc-AxiKw jIhbQX"
        >
          45
        </p>
        <span
          class="sc-AxhCb gbPKbT"
        >
          30 Last period
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Header fetches data and renders correctly 3`] = `<div />`;

exports[`Header shows an error message when fetch data with a failed status 1`] = `
<div>
  <div
    class="sc-AxmLO iyRexG"
  >
    <div
      class="sc-fzozJi csyUvS"
    >
      <div
        class="sc-AxhUy cuOvzs"
      >
        <label
          class="sc-AxgMl fxzWMk"
        >
          Filter by
        </label>
        <select
          class="sc-AxheI dXBtZm"
          data-testid="filter"
        >
          <option
            class="sc-Axmtr hBVTto"
            value="WEEK"
          >
            WEEK
          </option>
          <option
            class="sc-Axmtr hBVTto"
            value="MONTH"
          >
            MONTH
          </option>
          <option
            class="sc-Axmtr hBVTto"
            value="QUARTER"
          >
            QUARTER
          </option>
          <option
            class="sc-Axmtr hBVTto"
            value="YEAR"
          >
            YEAR
          </option>
        </select>
      </div>
    </div>
    <span
      class="sc-fzplWN jFvDHN"
      data-testid="error"
    >
      An error occurred when fetch the cards list, please try again.
    </span>
  </div>
</div>
`;

exports[`Header shows an error message when fetch data with a failed status 2`] = `<div />`;

exports[`Header shows an error message when promise gets rejected 1`] = `
<div>
  <div
    class="sc-AxmLO iyRexG"
  >
    <div
      class="sc-fzozJi csyUvS"
    >
      <div
        class="sc-AxhUy cuOvzs"
      >
        <label
          class="sc-AxgMl fxzWMk"
        >
          Filter by
        </label>
        <select
          class="sc-AxheI dXBtZm"
          data-testid="filter"
        >
          <option
            class="sc-Axmtr hBVTto"
            value="WEEK"
          >
            WEEK
          </option>
          <option
            class="sc-Axmtr hBVTto"
            value="MONTH"
          >
            MONTH
          </option>
          <option
            class="sc-Axmtr hBVTto"
            value="QUARTER"
          >
            QUARTER
          </option>
          <option
            class="sc-Axmtr hBVTto"
            value="YEAR"
          >
            YEAR
          </option>
        </select>
      </div>
    </div>
    <span
      class="sc-fzplWN jFvDHN"
      data-testid="error"
    >
      An error occurred when fetch the cards list, please try again.
    </span>
  </div>
</div>
`;

exports[`Header shows an error message when promise gets rejected 2`] = `<div />`;
